### PR TITLE
fix: crash when huge write

### PR DESCRIPTION
**Pull request by ZhongChaoqiang**
# Related-Issue
https://github.com/apache/incubator-pegasus/issues/903

# Change
when buffer of slog is more 2GB, then flush to slog file, it maybe crash.

### CODE REVIEW

**Review by foreverneverer — APPROVED** [review]

File: src/aio/disk_engine.cpp [hunk 1]
```diff
@@ -126,9 +126,9 @@ aio_task *disk_file::on_write_completed(aio_task *wk, void *ctx, error_code err,
         if (err == ERR_OK) {
             size_t this_size = (size_t)wk->get_aio_context()->buffer_size;
             dassert(size >= this_size,
```
Comment (acelyc111) [inline src/aio/disk_engine.cpp:128]: Better to use `dcheck_ge` to simplify code.

**Comment by acelyc111** [comment]
@ZhongChaoqiang The CI failed at https://github.com/XiaoMi/rdsn/runs/6189338284?check_suite_focus=true#step:8:1807, plz check it.

**Comment by ZhongChaoqiang** [comment]
ok, I’m checking it

从 Windows 版邮件<https://go.microsoft.com/fwlink/?LinkId=550986>发送

发件人: Yingchun ***@***.***>
发送时间: 2022年4月27日 17:22
收件人: ***@***.***>
抄送: Zhong ***@***.***>; ***@***.***>
主题: Re: [XiaoMi/rdsn] fix: crash when huge write (PR #1099)


@ZhongChaoqiang<https://github.com/ZhongChaoqiang> The CI failed at https://github.com/XiaoMi/rdsn/runs/6189338284?check_suite_focus=true#step:8:1807, plz check it.

―

**Comment by acelyc111** [comment]
@ZhongChaoqiang 
See details [here](https://github.com/XiaoMi/rdsn/runs/6189338415?check_suite_focus=true#step:8:1834), it tells you what's the problem is.
The variable in src/aio/disk_engine.cpp L269 is `uint32_t sz;`, but in src/aio/disk_engine.cpp L56, we want to dereference it as an uint64_t (`uint64_t &sz = *(uint64_t *)plength;`). So stack-buffer-overflow error occurs, and may cause following coredump.

File: src/aio/disk_engine.cpp (hunk 1 above)
Comment (ZhongChaoqiang) [reply src/aio/disk_engine.cpp]: done

**Comment by ZhongChaoqiang** [comment]
> @ZhongChaoqiang See details [here](https://github.com/XiaoMi/rdsn/runs/6189338415?check_suite_focus=true#step:8:1834), it tells you what the problem is. The variable in src/aio/disk_engine.cpp L269 is `uint32_t sz;`, but in src/aio/disk_engine.cpp L56, we want to dereference it as an uint64_t (`uint64_t &sz = *(uint64_t *)plength;`). So stack-buffer-overflow error occurs, and may cause following coredump.

Thanks very much

**Review by acelyc111 — APPROVED** [review]

**Comment by acelyc111** [comment]
Thanks for your contribution!
Could you describe how do you troubleshooting the bug you found when write with huge thoughput, and create a specify issue?

**Comment by xihong08** [comment]
> when to be written data size more 4GB,  it maybe happened on the server side in busying write status，crash  happened in aio_task::collapse  ::memcpy(dest, b.buffer, b.size);  this case resulting by memory overwriting .  and root cause is integer overflow，coz aio_context's buffer_size is 4 bytes,  in dsn::file::write_vector file_io.cpp L128 , the aio_context's buffer_size may be overflow when to be written data size more 4GB,  so crash it later.

**Comment by xihong08** [comment]
> Thanks for your contribution! Could you describe how do you troubleshooting the bug you found when write with huge thoughput, and create a specify issue?
    i can describe detailed case.   this maybe happened on the server side in busying write status，and crash happened in aio_task::collapse ::memcpy(dest, b.buffer, b.size);  it resulting by memory overwriting . and root cause is integer overflow，coz aio_context's buffer_size is 4 bytes, in dsn::file::write_vector file_io.cpp L128 , the aio_context's buffer_size may be overflow when to be written data size more 4GB, so crash it later.

**Review by foreverneverer — APPROVED** [review]

**Pull request merged**